java/memory/src/main/java/org/apache/arrow/memory/BaseAllocator.java

```diff
@@ -274,11 +314,14 @@ public ArrowBuf buffer(final int initialRequestSize, BufferManager manager) {
       return empty;
     }
 
-    // round to next largest power of two if we're within a chunk since that is how our allocator
-    // operates
-    final int actualRequestSize =
-        initialRequestSize < AllocationManager.CHUNK_SIZE ?
-          nextPowerOfTwo(initialRequestSize) : initialRequestSize;
+    // The actual buffer size may not equal the requested buffer size.
+    // This happens when the requested buffer size is within a chunk size. In that case, the actual buffer size
+    // must be a power of 2, and the specific rounding behavior depends on the rounding option:
+    // 1. If rounding up is used as the option, the actual buffer size will be requested size rounded up to the
+    // nearest power of 2. This is the default behavior.
+    // 2. If rounding down is used, the actual buffer size will be the requested size rounded down to the nearest
```

Comment:
I think it is this part that doesn't make sense to me.  Won't clients get an uexpected indexoutofboundsexception if they are expecting a request a buffer or a larger size but get one of the smaller size?

Reply:
@emkornfield Thanks for the good point. I think I understand what you are talking about. 

I think the fundamental reason here, is not that the user gets a smaller buffer unexpectedly. To get a smaller buffer, the user must have specified the rounding down option explicitly. 

I think the fundamental reason is that, the actual buffer size may differ from the requested buffer size, and neglecting this fact may cause problems. 

One of the problems is what you have mentioned in the comment (the problem of throwing IndexOutOfBoundException). Another problem is what I have discussed in the JIRA description (the problem of throwing OutOfMemoryException). 

Yet another problem is resource waste: For example, I am requesting a buffer of 10 MB, and by the default rounding up option, I get a buffer of 16 MB. Then I go ahead using the buffer, as if it has only 10 MB. This will waste the extra 6 MB memory space.

So if the user does not respect the fact that the requested and actual buffer sizes can be different, I guess there is no perfect solution for all these problems. 

However, if the user respects this fact, the perfect solution cannot be achieved either, without the options provide by this PR. So this PR makes the perfect solution possible.

Reply:
@liyafan82 thanks for the explanation.  The thing that seems strange to me is that this is different then the typical memory allocator abstraction.   For instance with malloc, at least as far as I'm aware you, don't have this level of control, it is up to the caller to make the decision on how much memory to ask for and the allocator can either satisfy it or it can't.  And different memory managers are better at worse and handling internal fragmentation which it seem like the problem here.

I agree it is surprising given the code that you shared, if you request a small amount of memory that appears to be within bounds to get an OOM unexpectedly.

As an alternative proposal, off the top of my head, what do you think of something like this API?

```
/**
 Determines if the allocation is likely to succeed (it accounts for 
 * for implementation details like padding buffers to certain lengths).
 */ 
bool canProbablyAllocate(int byteCount);
```
Then you can write the code in the JIRA as something like:
```
if (!allocator.canProbablyAllocate(requestSize)) {
   // Try making a smaller request size to avoid O
   requestSize = PreviousPowerOf2(requestSize);
}
buffer = allocator.allocate(requestSize);
```
The if statement could be placed into a static helper method someplace if it is a very common pattern.

It seems like we can then have different allocators that can have different internal fragmentation levels that make it more likely to succeed as long as there the request is within the different of memory allocated and their memory limit.

@praveenbingo thoughts?

Reply:
P.S. Even canProbablyAllocate seems like it might be superfluous, is there a reason you can't catch the OOM and make a smaller request size?  Performance?

Reply:
@emkornfield thanks a lot for your good proposal and good questions. 

You proposal aims at solving the rounding problem when the request size is smaller than the chunk size. However, we do not know the chunk size. AllocationManager#CHUNK_SIZE has package access level, and its value is platform dependent. This is the first problem we are facing.

Even if we can get the chunk size, say, by reflection. We are facing the second, maybe more important problem. The retry logic above depends on the buffer allocation policy of Arrow. To comply with the principle of encapsulation, I don't think the internal logic of Arrow should be exposed to external users, and the external code should rely on the internal features of Arrow. All the assumption it may have is that, the requested and actual buffer sizes can be different. 

This is because if one day, the allocation policy changes, all the user code would break. In fact, we are thinking about a proposal for alternative memory allocation policies, like the buffer size must be a multiple of some unit (e.g. 32 KB memory segment). Powers of 2 can cause too much waste for some scenarios. In this proposal, there is not a threshold like the chunk size, and the rounded size may not be a power of 2. 

So we think the retry logic above can be a good work around, but it may not be reliable in the long run. 

BTW, I am curious if it is possible for malloc that the actual buffer size is larger than the requested size? Is there a complex rule regulating the requested an actual buffer size?

Thank you in advance.

Reply:
@liyafan82 I don't think my proposal requires knowing chunk-size.  Just like growing memory by doubling the request each time, it seems reasonable to half your memory requests if you still think some is available but the allocator isn't cooperating.

Agree that other memory allocation policies might perform better.  In your example of 32KB memory segment, is it really worth rounding down if you happen to have only 31KB free and the request is 31.5KB?

Malloc can return more memory but this is transparent to the client (and the client can't make assumptions) about it, but the extra memory can be used to make things like Realloc if it is sufficiently large.

I'd like to think about this a little bit more.

Reply:
@siddharthteotia I think you've done the most recent work on this module, do you have thoughts on this API?

Reply:
@emkornfield Thanks a lot for your information about malloc.

I guess I need to explain a bit more about the 32 KB memory allocation policy. It means each Arrow buffer size must be rounded a multiple of 32 KB, like 64 KB, 96 KB, 1024 KB, etc. In the example you provided, if we request a 31.5 KB, and we have only 31 KB, we will first round the buffer size to 32 KB. Since there is no 32 KB memory, an OOM will be thrown. This will lead to a waste of 31 KB memory, but I think the waste can be much smaller than the power-of-2 policy.

I totally agree that your proposal works well even if the request size exceeds the chunk size. I think this "exponential back-off" method is a universal method that works for any cases. However, such back-off can be overly aggressive. For Arrow, it works well, because Arrow is also adopting a power-of-2 policy, which matches the exponential backing-off perfectly. In the 32 KB allocation policy, the back-off policy can be like this: given a request of 100 KB, we first try 128 KB, and if it fails, we try 96 KB. Given the amount of available memory (allocator.getLimit() - allocator.getAllocated()), this policy backs off more mildly and can quickly find the appropriate request size, without causing too much waster of memory. 

The key, is that the optimal backing-off policy depends on the internal allocation policy of Arrow, and such details should not be exposed to users. 

Please take your time and think about it, and thank you for all your innovative thoughts so far. 

@siddharthteotia, would you please give some comments?

Reply:
@liyafan82 I understand what you are saying, and I can see your point of view.   I just have an intuition that it might not be broadly applicable (since it isn't in available in other memory management systems I'm familiar with, for another example, the JVM's new doesn't provide this option).  This makes me think it might be optimizing at right level.  But I'm not an expert here.

I don't want to hold up this PR if others think it is a good idea to merge.  I just want to make sure we are considering each API we add because they are more difficult to remove once they are in place.

CC @praveenbingo again as well.

Reply:
@emkornfield , thanks for your effort in understanding our proposal. 

I agree with you that we should be careful when adding each new API. In addition, I believe in your expertise in memory management. So I do not want this PR to be merged, before we can reach an agreement. 

How about we solve this issue like this?

1. We remove the option of rounding up and rounding down from this PR, because we need to spend more time thinking about it. Maybe we can start a discussion in the community, and maybe we can find a better solution in the future.

2. We only preserve the option of providing the rounded buffer size. This is a relatively conservative solution, which does not affect the internal behavior of the allocator, and it will help users avoid OOM in their code. Given a buffer request, the users will know how much memory he/she is really asking for. 

What do you think? If you think it is OK, I will revise the code accordingly. 

Thank you in advance.

Reply:
@emkornfield @liyafan82 sorry for the delay.

I agree with @emkornfield suggestion, it makes for a cleaner API and has no surprises since the client is asking for exactly the amount of memory that is required and Arrow is not reducing it in anyway.
We can abstract the retry loop in an utility method if needed.

Reply:
@praveenbingo thank you for your comment. It seems rounding down the requested buffer size is not a good idea.

@praveenbingo and @emkornfield I have updated the PR and only preserve the logic for getting the rounded buffer size. Would you please help to see if it looks good to you?

Reply:
@liyafan82 I'm still concerned that the API might not be deterministic given other implementations of the API.  Again it might be OK, I just don't feel comfortable merging the change.  Could you please discuss on the ML?

Thank you.

Reply:
@emkornfield Sure. Thanks for the suggestion.
I will start a discussion on the ML.